### ISSUE TITLE
Questions about implementing layoutlmv3 objective II MIM

### ISSUE THREAD

**Issue by weicheng113** [microsoft/unilm#772]
Hi,

I have been using layoutlm series for some time and thanks for the sharing. I would like to pretrain layoutlmv3 in my domain. I am currently implementing the second objective Masked Image Modeling(MIM). I get a question regarding how image masking is implemented in MIM head? I can see BEiT uses mask_token parameter to mask out image tokens produced by PatchEmbed. LayoutLMv3Model does not have a mask_token parameter, as it does not need it after pretraining. I wonder if a modified version of LayoutLMv3Model(with mask_token included and applied) was used in MIM head when you implmented it? Could you please give me some tips and description how it is implemented, so that I can reference the idea.

With BEiT paper, I get a general idea on how to implement MIM head. I will generate image_masked_token_positions and labels(by Dall-e). Then I will use image_masked_token_positions to mask out image tokens produced by PatchEmbed(similar to BEiT). After that, image tokens and text tokens will be passed through transformer layers(the normal flow in LayoutLMv3Model).

Hi @HYPJUDY, this is Cheng again. Thanks for helping out my previous issue #751. If you have time, please help me also. Thanks.

Thanks,
Cheng

**Comment by HYPJUDY** [comment]
Thanks for your interest.
Please refer to LayoutLMv3's paper and BEiT's code to implement the MIM objective. Current LayoutLMv3Model is mainly released for fine-tuning downstream tasks, so it may not contain all parameters used in the pre-training phase. Therefore, you can add any necessary parameters for the implementation of MIM.

**Comment by weicheng113** [comment]
@HYPJUDY thanks for your response. I have implemented a version, but I am not sure if it is correct. I have a good reading of  LayoutLMv3's paper and LayoutLMv3Model implementation code. So far I have implemented two objectives, MLM(a slightly different version) and MIM. I feel it is a bit hard as there is nowhere I can check against(as the paper only contains ideas and formulas).

Thanks,
Cheng

**Comment by Rabona17** [comment]
@weicheng113 Hi Weicheng, I am also doing similar task, may I ask have you successfully recover the code for MIM loss? Did you retrain the dVAE from Beit?
Thanks,
Yuan

**Comment by weicheng113** [comment]
@Rabona17  I am not re-training aVAE. Using already trained dall-e as what was done in Beit.

**Comment by Rabona17** [comment]
@weicheng113 Hi Weicheng, thanks for the reply! Did you finally get some results from your own pretraining? Are they close to the paper's performance?
Thanks,
Yuan

**Comment by weicheng113** [comment]
@Rabona17 , No. I have only implemented 2 objectives out of 3. It was not easy, need to read relevant papers. My purpose was not to reproduce the paper result. I planned to train on DocBank or on my private documents.

**Comment by DiddyC** [comment]
Hey @weicheng113, I'm trying to implement the pre training objectives for LayoutlmV3 but the little information about it out there is very confusing. Is there any chance you'd be open to sharing your code?

**Comment by jmandivarapu1** [comment]
> @Rabona17 , No. I have only implemented 2 objectives out of 3. It was not easy, need to read relevant papers. My purpose was not to reproduce the paper result. I planned to train on DocBank or on my private documents. There are some other relevant similar papers, which have different objectives.

Hi @weicheng113 Can you provide your code for MLM implementation objective I

**Comment by 14H034160212** [comment]
Hi, for anyone who interested in the implementation of LayoutLMv3. Transfomers have updated the code for mask image modeling and the code is based on DEIT. You can inherit the code to implement the Mask Image Modeling for LayoutLMv3 and also you can also inherit the code from RoBERTa to implement the mask language modeling. For the word-patch alignment, I am still in progress. Free feel to have any discussion.
Here are the links:
[RoBERTa mask language modeling](https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling#robertabertdistilbert-and-masked-language-modeling)
[[DEIT mask image modeling example](https://github.com/huggingface/transformers/tree/main/examples/pytorch/image-pretraining#using-datasets-from--datasets)]

[More ideas for developing word patch alignment](https://github.com/NielsRogge/Transformers-Tutorials/issues/152)
